contracts/chain-adapters/Scroll_Adapter.sol

```diff
@@ -0,0 +1,106 @@
+     * @dev This message is marked payable because relaying the message will require
+     * a fee that needs to be propagated to the Scroll Bridge. It will not send msg.value
+     * to the target contract on L2.
+     * @param target L2 address to send message to.
+     * @param message Message to send to `target`.
+     */
+    function relayMessage(address target, bytes calldata message) external payable {
+        // We can specifically send a message with 0 value to the Scroll Bridge
+        // and it will not forward any ETH to the target contract on L2. However,
+        // we need to set the payable value to msg.value to ensure that the Scroll
+        // Bridge has enough gas to forward the message to L2.
+        l1ScrollMessenger.sendMessage{ value: _generateRelayerFee() }(target, 0, message, l2GasLimit);
+        emit MessageRelayed(target, message);
+    }
+
+    /**
+     * @notice Send `amount` of `l1Token` to `to` on Scroll. `l2Token` is the Scroll address equivalent of `l1Token`.
+     * @dev This method is marked payable because relaying the message might require a fee
+     * to be paid by the sender to forward the message to L2. However, it will not send msg.value
+     * to the target contract on L2.
+     * @param l1Token L1 token to bridge.
+     * @param l2Token L2 token to receive.
+     * @param amount Amount of `l1Token` to bridge.
+     * @param to Bridge recipient.
+     */
+    function relayTokens(
+        address l1Token,
+        address l2Token,
+        uint256 amount,
+        address to
+    ) external payable {
+        IL1GatewayRouter _l1GatewayRouter = l1GatewayRouter;
+
+        // Confirm that the l2Token that we're trying to send is the correct counterpart
+        // address
+        address _l2Token = _l1GatewayRouter.getL2ERC20Address(l1Token);
+        require(_l2Token == l2Token, "l2Token Mismatch");
+
+        // Bump the allowance
+        IERC20(l1Token).safeIncreaseAllowance(address(_l1GatewayRouter), amount);
+
+        // The scroll bridge handles arbitrary ERC20 tokens and is mindful of
+        // the official WETH address on-chain. We don't need to do anything specific
+        // to differentiate between WETH and a separate ERC20.
+        // Note: This happens due to the L1GatewayRouter.getERC20Gateway() call
+        // Note: dev docs: https://docs.scroll.io/en/developers/l1-and-l2-bridging/eth-and-erc20-token-bridge/
+        _l1GatewayRouter.depositERC20{ value: _generateRelayerFee() }(l1Token, to, amount, l2GasLimit);
+        emit TokensRelayed(l1Token, l2Token, amount, to);
+    }
+
+    /**************************************
+     *         INTERNAL FUNCTIONS         *
+     **************************************/
+
+    /**
+     * @notice Generates the relayer fee for a message to be sent to L2.
+     */
+    function _generateRelayerFee() internal view returns (uint256) {
+        return l2GasPriceOracle.estimateCrossDomainMessageFee(l2GasLimit);
```

Comment:
Do we need a buffer on this at all?

Reply:
Do you mean like a gas padding? We could do that - the L2 price oracle is updated by a scroll bot so it may be less than accurate during high network activity.

Do you think we should add that into the code and initialize it to zero so that if we need it, we can use it

Reply:
fee padding, not just gas-limit padding. I guess the `gasLimit` value can be used as a buffer, right?

Reply:
Sorry, that's what I meant. The value we're propagating in the `{ value: _generateRelayerFee() }`